yarn-project/p2p/src/service/data_store.ts

```diff
@@ -0,0 +1,259 @@
+import type { AztecKVStore, AztecMap } from '@aztec/kv-store';
+
+import { type Batch, type Datastore, Key, type KeyQuery, type Pair, type Query } from 'interface-datastore';
+import type { AwaitIterable } from 'interface-store';
+import filter from 'it-filter';
+import map from 'it-map';
+import sort from 'it-sort';
+import take from 'it-take';
+
+type MemoryItem = {
+  lastAccessedMs: number;
+  data: Uint8Array;
+};
+
+type BatchOp = {
+  type: 'put' | 'del';
+  key: Key;
+  value?: Uint8Array;
+};
+
+export class AztecDatastore implements Datastore {
+  #memoryDatastore: Map<string, MemoryItem>;
+  #dbDatastore: AztecMap<string, Uint8Array>;
+  #dirtyItems = new Set<string>();
+
+  #batchOps: BatchOp[] = [];
+
+  // private dirtyItems = new Set<string>();
+  private threshold: number;
+  private maxMemoryItems: number;
+
+  constructor(db: AztecKVStore, { maxMemoryItems, threshold } = { maxMemoryItems: 50, threshold: 5 }) {
+    this.#memoryDatastore = new Map();
+    this.#dbDatastore = db.openMap('p2p_datastore');
+
+    this.maxMemoryItems = maxMemoryItems;
+    this.threshold = threshold;
+  }
+
+  has(key: Key): boolean {
+    return this.#memoryDatastore.has(key.toString()) || this.#dbDatastore.has(key.toString());
+  }
+
+  async get(key: Key): Promise<Uint8Array> {
+    // console.log('getting data', key);
+    const keyStr = key.toString();
+    const memoryItem = this.#memoryDatastore.get(keyStr);
+    // console.log('res', memoryItem);
+    if (memoryItem) {
+      memoryItem.lastAccessedMs = Date.now();
+      return memoryItem.data;
+    }
+    const dbItem = this.#dbDatastore.get(key.toString());
+
+    if (!dbItem) {
+      throw new Error(`Key not found: ${key.toString()}`);
+    }
+
+    // don't call this._memoryDatastore.set directly
+    // we want to get through prune() logic with fromDb as true
+    await this._put(key, dbItem, true);
+
+    return dbItem;
+  }
+
+  private async _put(key: Key, val: Uint8Array, fromDb = false): Promise<Key> {
```

Comment:
Can you restructure this so that the public methods are at the top and private underneath?